src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -233,31 +175,74 @@ private unsafe SocketError ProcessIOCPResult(bool success, int bytesTransferred,
+                    _pendingOverlappedForCancellation = overlapped;
+                    _registrationToCancelPendingIO = cancellationToken.UnsafeRegister(static s =>
+                    {
+                        // Try to cancel the I/O.  We ignore the return value (other than for logging), as cancellation
+                        // is opportunistic and we don't want to fail the operation because we couldn't cancel it.
+                        var thisRef = (SocketAsyncEventArgs)s!;
+                        SafeSocketHandle handle = thisRef._currentSocket!.SafeHandle;
+                        if (!handle.IsClosed)
+                        {
+                            try
+                            {
+                                bool canceled = Interop.Kernel32.CancelIoEx(handle, thisRef._pendingOverlappedForCancellation);
+                                if (NetEventSource.Log.IsEnabled())
+                                {
+                                    NetEventSource.Info(thisRef, canceled ?
+                                        "Socket operation canceled." :
+                                        $"CancelIoEx failed with error '{Marshal.GetLastWin32Error()}'.");
+                                }
+                            }
+                            catch (ObjectDisposedException)
+                            {
+                                // Ignore errors resulting from the SafeHandle being closed concurrently.
+                            }
+                        }
+                    }, this);
+                }
+                if (!bufferToPin.Equals(default))
+                {
+                    _singleBufferHandle = bufferToPin.Pin();
+                }
 
-                _singleBufferHandle = bufferToPin.Pin();
+                // We've finished setting up and launching the operation.  Coordinate with the callback.
+                ulong packedResult = Interlocked.Exchange(ref _asyncCompletionOwnership, 1);
+                if (packedResult == 0)
+                {
+                    // We won the race condition with the callback. It now owns completion and clean up.
+                    return SocketError.IOPending;
+                }
 
-                _asyncProcessingState = AsyncProcessingState.Set;
-            }
-            else
-            {
-                _asyncProcessingState = AsyncProcessingState.None;
-                FinishOperationSync(socketError, bytesTransferred, SocketFlags.None);
+                // The callback was already invoked and transferred ownership to us, so now behave as if the operation completed synchronously.
+                // Since the success/bytesTransferred arguments passed into this method are stale, we need to retrieve the actual status info
+                // from the overlapped directly.  It's also now our responsibility to clean up as GetIOCPResult would have, so free the overlapped.
+                FreeNativeOverlapped(overlapped);
```

Comment:
Why we are not also cleaning up `_singleBufferHandle` and `_registrationToCancelPendingIO` here? `CleanupIOCPResult()` seems to be called only if the ownership was transferred to the callback...

Reply:
> CleanupIOCPResult() seems to be called only if the ownership was transferred to the callback...

It's called by Complete() which is invoked at the end of every operation, whether it completes synchronously or asynchronously. At least that's the intention... do you see otherwise?

Reply:
Ah, I got confused by "synchronous" vs "as-if-synchronous" terminology in `CleanupIOCPResult` comment. I see. If it's truly synchronous, `_asyncCompletionOwnership` will remain 0 and buffer will not be pinned, and only then `CleanupIOCPResult` will not be called on `Complete`. And if it was asynchronous, the flag will be 1 (not 0) if it's "as-if-synchronous", and bit-magicked-info-with-top-bit-set (not 0) if "slow" asynchronous. So `CleanupIOCPResult` will be called. So it should be ok :)